BDR-3567 - Add http health check information (#4132)
EnterpriseDB/docs: product_docs/docs/pgd/5/routing/proxy.mdx

Having said that, most of the time, upon loss of the current Raft leader the new Raft leader gets elected by BDR within a few seconds provided more than half of the nodes (quorum) are still up. Hence, if the Raft leader is down but the write leader is still up then proxy can be configured to allow routing by keeping `consensus_grace_period` to a non-zero, positive value. The proxy will wait for the Raft leader to get elected during this period before stopping routing. This may be helpful in some cases where availability is more important.

> ### Proxy health check

Reviewer: Move this section to end of this page, under `Managing PGD Proxy`, parallel to [Proxy log location](https://www.enterprisedb.com/docs/pgd/latest/routing/proxy/#proxy-log-location). We don't want to highlight it and also probes are more about Managing Proxy.

> PGD Proxy provides the following HTTP(s) health check API endpoints. These APIs are `GET` requests and do not require a request body.

Reviewer: add blank line after this line

Author: done

```

> GET /health/is-live

Reviewer: may be first add is-ready

GET /health/is-ready
```

> #### Liveness

Reviewer: let's move it after readiness... this is not important endpoint and may unnecessarily raise question

Author: done

> Liveness check either return `true` with HTTP status code `200 (OK)` or an error but never `false`. This is due to the fact that Proxy service exits immediately effectively stopping HTTP server if it finds a critical error on startup that cannot be recovered or prevents further operations.

Reviewer: Need to add one more stmt in the end which is the actual reason why it never returns false. - `Once started proxy works in self healing mode, it tries to recover from any runtime error...` (not necessarily in same language)

#### Readiness

On receiving this request, proxy checks if it can successfully route traffic through the `proxy.endpoint`. If the check returns successfully then API returns `true` with HTTP status code `200 (OK)` else `false` with HTTP status code `500 (Internal Server Error)`. It will return `false` in the following scenarios.
* Proxy detects that the Raft leader is not available and `consensus_grace_period` has expired.
* Proxy detects that the write leader is not available.
* Proxy cannot establish a TCP connection to the current write leader
* The TLS certificate on the current write leader is expired

It is important to know that once started proxy always remains in available state even if this check return false because the proxy can recover itself automatically (self heal) once the BDR cluster or connectivity is restored.

!!! Note
    Proxy will perform the readiness check against the backend Postgres only when the readiness endpoint is requested and not on periodic basis.

> #### Configuration

Reviewer: Move this section under [Configuration](https://www.enterprisedb.com/docs/pgd/latest/routing/proxy/#proxy-log-location), after [PGD proxy user](https://www.enterprisedb.com/docs/pgd/latest/routing/proxy/#proxy-log-location) section and before PGD Proxy service. Saying only required if health check probes are required in environment like k8s.

On this page, here, only add link to that configuration section saying see this for how to enable these endpoints etc. 

Similarly, same link to this page from Configuration section.

Author: done

By default the proxy health check is disabled. It can be enabled by adding the config `proxy.http.enable: true` in the `pgd-proxy-config.yml` file as shown below.
```
  proxy:
	  name: proxy-a1
	  endpoint: "host=proxy-a1 port=6432 dbname=bdrdb user=pgdproxy "
	  http:
  	    enable: true
```

> Default HTTP server address is `:8080` with timeout `10s`. Below is a sample config to set the custom values of host, port and other configs.

Reviewer: move this first line before default sample config

  proxy:
	  name: proxy-a1
	  endpoint: "host=proxy-a1 port=6432 dbname=bdrdb user=pgdproxy "
	  http:
  	    enable: true
  	    host: "0.0.0.0"
  	    port: 8080
        secure: false
        cert_file: "/etc/tpa/server.crt"
        key_file: "/etc/tpa/server.key"
	    probes:
  	      timeout: 10s
```

The `proxy.endpoint` parameter is used by readiness API endpoint which try to connect to the proxy server using this endpoint. Host and port given in the endpoint should be of proxy (e.g. host=proxy-a1 port=6432). Please note this is different from cluster endpoints where host and port are of the backend BDR server. User and database could be the same as per setup though. Proxy user (`pgdproxy` in above example) should have the database login access and its password set up in `~/.pgpass` file. If cluster is setup through TPAExec, it creates the `pgdproxy` user as an OS user as well as database user with `bdr_superuser` role and have it password set in the `pgpass` file.

To enable HTTPS set the config parameter `proxy.http.secure: true`. If enabled, the `cert_file` and `key_file` config parameters become mandatory and expect server certificate and key file paths respectively.

> The `proxy.http.probes.timeout` is used by readiness API endpoint. Readiness probe will return `false` if proxy does not receive the response from backend Postgres within this time period. This value should be set based on failover time and retry attempts to avoid false positives.

Reviewer: `from backend Postgres` -> `from current write leader`

Author: done

See [Connection management](../routing) for more information on the PGD side of configuration and management of PGD Proxy.

### Proxy health check

> PGD Proxy provides the following HTTP(s) health check API endpoints. These APIs are `GET` requests and do not require a request body. See [Configurations](installing_proxy#configuring-health-check) for more details about enabling and configuring this.

Reviewer 2: ```suggestion
PGD Proxy provides the following HTTP(s) health check API endpoints. The API endpoints responds to `GET` requests. The endpoints need to be enabled and configured before use; see [Configurations](installing_proxy#configuring-health-check) for how to do this.
```

Author: Sure. Thanks

```
GET /health/is-ready
GET /health/is-live
```

#### Readiness

> On receiving valid request, proxy checks if it can successfully route connections to the current write leader. If the check returns successfully then API returns `true` with HTTP status code `200 (OK)` else `false` with HTTP status code `500 (Internal Server Error)`.

Reviewer 2: What makes a invalid request? "On receiving a request, the proxy..."

And is that 'true' because 200 was returned or is there a body containing 'true' returned? It reads like both things happen.

Reviewer 2: ```suggestion
On receiving valid 'GET' request, the proxy checks if it can successfully route connections to the current write leader. If the check returns successfully then API gives a response with a body containing `true` and a HTTP status code `200 (OK)`. Otherwise, it returns a body containing `false` with the HTTP status code `500 (Internal Server Error)`.
```

Author: Sounds good. Changing it.

#### Liveness

> Liveness check either return `true` with HTTP status code `200 (OK)` or an error but never `false`. This is because the HTTP server listening for request is automatically stopped if the Proxy service fails to start or exists.

Reviewer 2: Don't you mean the request only ever returns 'true'; if the request cannot be made, it is because the http server has been stopped because the Proxy service has not started or stopped. 

"fails to start or exists" is contradictory.

Author: Yes, that's what it means. I used the word "**exit**"  because Proxy can be stopped gracefully or exit abruptly due to some critical error conditions that cannot be handled by the proxy

Reviewer 2: Ah, exit.... So should be "fails to start or exits" (not exists).

See [Connection management](../routing) for more information on the PGD side of configuration and management of PGD Proxy.

### Proxy health check

PGD Proxy provides the following HTTP(s) health check API endpoints. The API endpoints responds to `GET` requests. The endpoints need to be enabled and configured before use; see [Configurations](installing_proxy#configuring-health-check) for how to do this.

```
GET /health/is-ready
GET /health/is-live
```

#### Readiness

On receiving valid 'GET' request, the proxy checks if it can successfully route connections to the current write leader. If the check returns successfully then API gives a response with a body containing `true` and a HTTP status code `200 (OK)`. Otherwise, it returns a body containing `false` with the HTTP status code `500 (Internal Server Error)`.

#### Liveness

> Liveness check either return `true` with HTTP status code `200 (OK)` or an error but never `false`. This is because the HTTP server listening for request is automatically stopped if the Proxy service fails to start or exits.

Reviewer 2: ```suggestion
The Liveness check can either return `true` with HTTP status code `200 (OK)` or a timeout error but never `false`. This is because the HTTP server listening for request is automatically stopped if the Proxy service fails to start or exits.
```
